src/querystring.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{error::Error, Context};
+use http::StatusCode;
+use serde::Deserialize;
+
+/// An extension trait for `Context`, providing query string deserialization.
+pub trait ExtractQuery<'de> {
+    fn query<T: Deserialize<'de>>(&'de self) -> Result<T, Error>;
+}
+
+impl<'de, Data> ExtractQuery<'de> for Context<Data> {
+    fn query<T: Deserialize<'de>>(&'de self) -> Result<T, Error> {
+        let querystr = self.uri().query();
+
+        if querystr.is_none() {
+            return Err(Error::from(StatusCode::BAD_REQUEST));
+        }
+
+        let query = serde_urlencoded::from_str(querystr.unwrap())
+            .map_err(|_| Error::from(StatusCode::INTERNAL_SERVER_ERROR))?;
+
+        Ok(query)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::executor::block_on;
+    use http_service::Body;
+    use http_service_mock::make_server;
+    use serde_derive::Deserialize;
+
+    #[derive(Deserialize)]
+    struct Params {
+        msg: String,
+    }
+
+    async fn handler(cx: crate::Context<()>) -> Result<String, Error> {
+        let p = cx.query::<Params>()?;
+        Ok(p.msg)
+    }
+
+    fn app() -> crate::App<()> {
+        let mut app = crate::App::new(());
+        app.at("/").get(handler);
+        app
+    }
+
+    #[test]
+    fn successfully_deserialize_query() {
+        let app = app();
+        let mut server = make_server(app.into_http_service()).unwrap();
+        let req = http::Request::get("/?msg=Hello")
+            .body(Body::empty())
+            .unwrap();
+        let res = server.simulate(req).unwrap();
+        assert_eq!(res.status(), 200);
+        let body = block_on(res.into_body().into_vec()).unwrap();
+        assert_eq!(&*body, &*b"Hello");
+    }
+
+    #[test]
+    fn unsuccessfully_deserialize_query() {
+        let app = app();
+        let mut server = make_server(app.into_http_service()).unwrap();
+        let req = http::Request::get("/").body(Body::empty()).unwrap();
+        let res = server.simulate(req).unwrap();
+        assert_eq!(res.status(), 400);
+    }
+}
```

Comment:
Could we perhaps also include a test for a malformed query string? E.g. passing `?error=should_fail` or something similar, where `msg=Hello` would be expected.

Reply:
Done. I should've caught this, but it's always good to have a reviewer. Thanks!